### PR TITLE
fix: add integration tests for (user) organization repositories

### PR DESCRIPTION
**Pull request by iamacook**
## Summary

We test the (user) organization domain via controller tests. This adds integration tests for the repositories of the aforementioned controller, covering database-relevant cases.

## Changes

- Add test coverage for (user) organization repositories
- Simplify interfaces that have unnecessary argumennts
- Tweak small edge cases found by tests

### CODE REVIEW

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 1]
```diff
@@ -0,0 +1,1414 @@
...
+        {
+          createdAt: expect.any(Date),
+          id: expect.any(Number),
+          status: 'ACTIVE', // Not pending
+          updatedAt: expect.any(Date),
+        },
+      ]);
+    });
+
+    it('should throw an error if the signer_address does not exist', async () => {
```
Comment (PooyaRaki) [inline src/domain/users/user-organizations.repository.integration.spec.ts:592]: For visibility, We have decided to take the authentication logic out of the repository, but it will happen later.

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 2]
```diff
@@ -0,0 +1,1414 @@
...
+      await dbWalletRepo.insert({
+        user: owner.generatedMaps[0],
+        address: authPayloadDto.signer_address,
+      });
+      const org = await dbOrgRepo.insert({
+        name: orgName,
+        status: 'ACTIVE',
+      });
+      const orgId = org.generatedMaps[0].id;
+      await dbUserOrgRepo.insert({
```
Comment (PooyaRaki) [inline src/domain/users/user-organizations.repository.integration.spec.ts:676]: Nit: the tests checks what it should but logically it would be better to have an active admin and check against a second pending admin because it's not possible to have an organization without an active admin.

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 3]
```diff
@@ -0,0 +1,1414 @@
...
+        name: orgName,
+        status: 'ACTIVE',
+      });
+      const userOrgRole = faker.helpers.arrayElement(UserOrgRoleKeys);
+      const orgId = org.generatedMaps[0].id;
+      const userOrg = await dbUserOrgRepo.insert({
+        user: user.generatedMaps[0],
+        organization: org.generatedMaps[0],
+        role: userOrgRole,
+        status: 'INVITED',
```
Comment (PooyaRaki) [inline src/domain/users/user-organizations.repository.integration.spec.ts:719]: Nit: the tests checks what it should but logically it would be better to have an active admin and check against a second pending admin because it's not possible to have an organization without an active admin.

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 4]
```diff
@@ -0,0 +1,1414 @@
...
+
+      await expect(
+        userOrgRepo.acceptInvite({
+          authPayload: new AuthPayload(authPayloadDto),
+          orgId,
+        }),
+      ).rejects.toThrow('Organization not found.');
+    });
+
+    it('should throw an error if the user organization is not INVITED', async () => {
```
Comment (PooyaRaki) [inline src/domain/users/user-organizations.repository.integration.spec.ts:806]:
```ts
    it('Should throw an error if the user is already a member of the organization', async () => {
```

another test:
```ts
it('Should throw an error if the user is not INVITED to the organization', async () => {
````

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 5]
```diff
@@ -0,0 +1,1414 @@
...
+        updatedAt: expect.any(Date),
+      });
+      await expect(
+        dbUserRepo.findOneOrFail({
+          where: { id: user.generatedMaps[0].id },
+        }),
+      ).resolves.toEqual({
+        createdAt: expect.any(Date),
+        id: userId,
+        status: 'PENDING', // Remains PENDING
```
Comment (PooyaRaki) [inline src/domain/users/user-organizations.repository.integration.spec.ts:888]: I'd expect the user to become active once he has accepted/declined an invitation, it makes sense to activate a user once they have interaction with the app. We can fix it in another PR if necessary.

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 6]
```diff
@@ -0,0 +1,1414 @@
...
+
+      await expect(
+        userOrgRepo.declineInvite({
+          authPayload: new AuthPayload(authPayloadDto),
+          orgId,
+        }),
+      ).rejects.toThrow('Organization not found.');
+    });
+
+    it('should throw an error if the user organization is not INVITED', async () => {
```
Comment (PooyaRaki) [inline src/domain/users/user-organizations.repository.integration.spec.ts:947]:
```ts
    it('Should throw an error if the user is already a member of the organization', async () => {
```

another test:
```ts
it('Should throw an error if the user is not INVITED to the organization', async () => {
````

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 7]
```diff
@@ -0,0 +1,1414 @@
...
+        userOrgRepo.updateRole({
+          authPayload: new AuthPayload(authPayloadDto),
+          orgId,
+          role: 'ADMIN',
+          userId,
+        }),
+      ).rejects.toThrow('User not found.');
+    });
+
+    it('should throw an error if the organization has no ACTIVE ADMINs', async () => {
```
Comment (PooyaRaki) [inline src/domain/users/user-organizations.repository.integration.spec.ts:1176]:
```suggestion
    it('should throw an error if user does not have access to upgrade to ADMIN', async () => {
```
An organization cannot exist without an admin, so I would test the scenarios that currently occur in our system.

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 8]
```diff
@@ -0,0 +1,1414 @@
...
+          authPayload: new AuthPayload(authPayloadDto),
+          orgId,
+          role: 'MEMBER',
+          userId,
+        }),
+      ).rejects.toThrow('Cannot remove last admin.');
+    });
+  });
+
+  describe('removeUser', () => {
```
Comment (PooyaRaki) [inline src/domain/users/user-organizations.repository.integration.spec.ts:1256]: Since Authentication/Authorization logic happens in the repo, we should add some tests to make sure an unauthorized user cannot remove members of an org.
The same applies to other methods like `updateRole`, `acceptInvite`, `declineInvite`, `inviteUser` but I won't add a comment on each method to keep the review concise.

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 9]
```diff
@@ -0,0 +1,1414 @@
...
+      await expect(
+        userOrgRepo.removeUser({
+          authPayload: new AuthPayload(authPayloadDto),
+          orgId,
+          userId,
+        }),
+      ).rejects.toThrow('User not found.');
+    });
+
+    it('should throw an error if the organization has no ACTIVE ADMINs', async () => {
```
Comment (PooyaRaki) [inline src/domain/users/user-organizations.repository.integration.spec.ts:1350]: Is it possible to have an org without an active admin? If not, I'd remove this test.

**Comment by PooyaRaki** [comment]
@iamacook I accidentally approved the PR. I won’t dismiss the review to ensure the comments remain visible.

File: src/domain/organizations/organizations.repository.spec.ts [hunk 10]
```diff
@@ -0,0 +1,670 @@
...
+      const updatedAtTime = updatedAt.getTime();
+
+      expect(createdAtTime).toBeGreaterThanOrEqual(before);
+      expect(createdAtTime).toBeLessThanOrEqual(after);
+
+      expect(updatedAtTime).toBeGreaterThanOrEqual(before);
+      expect(updatedAtTime).toBeLessThanOrEqual(after);
+    });
+
+    it('should update updatedAt when updating a User', async () => {
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/domain/organizations/organizations.repository.spec.ts]: Updated in d556d8e14bac3e535f72e269f55bb37a7031d792.

File: src/domain/organizations/organizations.repository.spec.ts [hunk 11]
```diff
@@ -0,0 +1,670 @@
...
+      });
+
+      const dbUser = await dbUserRepo.findOneOrFail({
+        where: { id: userId },
+      });
+      const dbOrg = await dbOrgRepo.findOneOrFail({
+        where: { id: org.id },
+      });
+      const dbUserOrg = await dbUserOrgRepo.findOneOrFail({
+        where: { user: { id: userId } },
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/domain/organizations/organizations.repository.spec.ts]: Adjusted in d556d8e14bac3e535f72e269f55bb37a7031d792.

File: src/domain/organizations/organizations.repository.integration.spec.ts [hunk 12]
```diff
@@ -0,0 +1,666 @@
...
+      });
+
+      const dbUser = await dbUserRepo.findOneOrFail({
+        where: { id: userId },
+      });
+      const dbOrg = await dbOrgRepo.findOneOrFail({
+        where: { id: org.id },
+      });
+      const dbUserOrg = await dbUserOrgRepo.findOneOrFail({
+        where: { user: { id: userId } },
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/domain/organizations/organizations.repository.integration.spec.ts:195]: Adjusted in d556d8e14bac3e535f72e269f55bb37a7031d792.

File: src/domain/organizations/organizations.repository.integration.spec.ts [hunk 13]
```diff
@@ -0,0 +1,666 @@
...
+        orgRepo.create({
+          userId,
+          name: orgName,
+          status: orgStatus,
+        }),
+      ).rejects.toThrow(
+        'null value in column "status" of relation "users" violates not-null constraint',
+      );
+
+      await expect(dbUserRepo.find()).resolves.toEqual([]);
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/domain/organizations/organizations.repository.integration.spec.ts:260]: I split it in d556d8e14bac3e535f72e269f55bb37a7031d792, with the suggested test being a `todo`. I will address it shortly.

File: src/domain/organizations/organizations.repository.integration.spec.ts [hunk 14]
```diff
@@ -0,0 +1,666 @@
...
+      const org = await orgRepo.create({
+        userId,
+        name: orgName,
+        status: orgStatus,
+      });
+
+      await orgRepo.delete(org.id);
+
+      await expect(
+        dbOrgRepo.findOne({ where: { id: org.id } }),
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/domain/organizations/organizations.repository.integration.spec.ts]: Changed in d556d8e14bac3e535f72e269f55bb37a7031d792.

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 15]
```diff
@@ -0,0 +1,1414 @@
...
+      expect(createdAt).toEqual(updatedAt);
+
+      expect(createdAt.getTime()).toBeGreaterThanOrEqual(before);
+      expect(createdAt.getTime()).toBeLessThanOrEqual(after);
+
+      expect(updatedAt.getTime()).toBeGreaterThanOrEqual(before);
+      expect(updatedAt.getTime()).toBeLessThanOrEqual(after);
+    });
+
+    it('should update updatedAt when updating a User', async () => {
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts]: Changed in d556d8e14bac3e535f72e269f55bb37a7031d792.

File: src/domain/users/user-organizations.repository.integration.spec.ts [hunk 16]
```diff
@@ -0,0 +1,1414 @@
...
+          return {
+            userId: expect.any(Number),
+            orgId,
+            role: user.role,
+            status: 'INVITED',
+          };
+        }),
+      );
+
+      const walletsInDb = await dbWalletRepo.find({
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts]: Adjusted in d556d8e14bac3e535f72e269f55bb37a7031d792.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 2 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts:662]: Changed in d556d8e14bac3e535f72e269f55bb37a7031d792.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 3 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts:711]: In this case, an invitee needn't be an `ADMIN` to accept/decline an invite. I scaffolded tests in d556d8e14bac3e535f72e269f55bb37a7031d792 to tests non-invitees that I will shortly complete.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 4 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts]: I scaffolded the new test in d556d8e14bac3e535f72e269f55bb37a7031d792 and will complete it shortly.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 6 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts]: I scaffolded the new test in d556d8e14bac3e535f72e269f55bb37a7031d792 and will complete it shortly.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 7 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts]: Changed in d556d8e14bac3e535f72e269f55bb37a7031d792.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 8 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts:1262]: I scaffolded new tests in d556d8e14bac3e535f72e269f55bb37a7031d792 that I will complete shortly.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 9 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts]: It is not. I removed the test in d556d8e14bac3e535f72e269f55bb37a7031d792.

File: src/domain/organizations/organizations.repository.integration.spec.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (iamacook) [reply src/domain/organizations/organizations.repository.integration.spec.ts:260]: Added the test in df962dc640c99b5d2584271692f502637bff28e0.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 3 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts:711]: Added the tests in df962dc640c99b5d2584271692f502637bff28e0.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 4 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts]: Added the test in df962dc640c99b5d2584271692f502637bff28e0.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 8 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts:1463]: Added the tests in df962dc640c99b5d2584271692f502637bff28e0.

Comment (PooyaRaki) [reply src/domain/users/user-organizations.repository.integration.spec.ts:1463]: `inviteUsers` is still missing `Authorization` tests.
nit: `updateRole`:
`Should throw an error if user does not have access to upgrade other members' roles`

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 2 above)
Comment (PooyaRaki) [reply src/domain/users/user-organizations.repository.integration.spec.ts:662]: Again the concern was an org not having an active admin, but since you added many more tests to check different scenarios I'll resolve it. Thanks for adding new tests.

File: src/domain/users/user-organizations.repository.integration.spec.ts (hunk 8 above)
Comment (iamacook) [reply src/domain/users/user-organizations.repository.integration.spec.ts:1602]: I added relevant tests in cf92c90a239f7c658457c7765f2076c1f25bd9db.

**Review by PooyaRaki — APPROVED** [review]
LGTM. Great work @iamacook
